### ISSUE TITLE
Rename `Transaction.{input, output}` to plural?

### ISSUE THREAD

**Issue by Kixunil** [rust-bitcoin/rust-bitcoin#822]
Using singular sounds weird for naming repeated data. Plural is used in these situations pretty much everywhere I've seen, including this crate (e.g. in PSBT). This API-breaking that would affect multiple dependents but it's trivial to fix and I think it's worth it.

**Comment by apoelstra** [comment]
I prefer the current "`input[i]` refers to the 'i'th input" situation, but googling for this particular holy war shows I am in the minority, and I have no strong feelings about this.

**Comment by Kixunil** [comment]
Interesting view. I found things like `tx.input.iter()...` quite common which is not that appealing to me.

**Comment by apoelstra** [comment]
Yeah, that's fair. My view comes from C, and in Rust it's much more common to treat vectors as containers instead of constantly indexing into them.

**Comment by tcharding** [comment]
If we are going to do this should we do it after we move `Transaction` to `primitives`?

**Comment by Kixunil** [comment]
I think it doesn't matter but I want it done before any `primitives` release.

**Comment by apoelstra** [comment]
I think we should leave the existing names alone. This is a case of #3166.

We could add accessors with plural names and try to encourage people to use those. (If `primitives 2.0` gets rid of the public fields that might even make the transition easier.)

**Comment by Kixunil** [comment]
Oh, maybe we can do the same trick as `cargo_toml` did! Because of the `Cargo.toml` format change they had to change the type of a public field. So to make transition easier they added a method returning the real value in an older version and deprecated the field, then they broke it in the next version keeping the method. The result being one can use ranged version specification. However using it from libraries may be brittle if they expose the type in the API too. But in that case we should be able to trivially semver-trick the leaf types and implement a likely-zero-cost conversions by just moving fields around (if the layouts end up being the same the conversion is zero cost and this is likely to happen).

**Comment by Kixunil** [comment]
> I think we should leave the existing names alone. This is a case of #3166.

I really, really, really don't like this. `for input in self.input` looks totally like a bug or bad variable naming unless you are familiar with rust-bitcoin.

**Comment by apoelstra** [comment]
We should add `self.inputs()` then. We could even deprecate the field if you want to discourage direct accesses. But I think we should keep the field (deprecation and all) around even for primitives 1.0 to minimize gratuitous code breaking.

**Comment by Kixunil** [comment]
One reason to rename the field is to avoid annoying difference between `inputs()` and `inputs_mut()`.

So here's the suggestion: add the accessors and deprecate the field ASAP with a message, in addition to suggestion, asking for feedback in a specially prepared GH issue. Then we can decide if we should rename them or not based on the feedback. We will need to give it enough time though. We could also have primitives 2.0 one day but I hope we will not.

**Comment by apoelstra** [comment]
Sounds good.

**Comment by tcharding** [comment]
I just had the second attempt at this I've made, it opens such a can of worms. No clue what to do about this.

**Comment by apoelstra** [comment]
I think we should add `inputs()` and `inputs_mut()` accessors and call it a day. People who have strong feelings about shit like this can just use those.

If you are feeling really ambitious you can try deprecating the fields, but I think that just won't work for people (due to borrowck limitations discussed above) and as a result we'll have to un-deprecate them.

**Comment by Kixunil** [comment]
I had an idea. IIRC `rustc` has Json output for error messages. I think that if we (I 🙂 ) hack up something to find all spans of "no such field output, help, there's outputs" and replace them we could employ this strategy:

0. Test this idea a bit but do not merge yet
1. Continue working on `primitives` as if we didn't plan this
2. Once we're comfortable with releasing `primitives` release a special crate with version 1.0-transition and all the crates depending on it
3. Change the field names and run the tool to fix up everything automatically
4. Release 1.0 while very visibly telling people to first set their version to the transition version, fixing all compile errors (unrelated to fields change) manually, then change it again and run our tool

Do you think it'd be acceptable if I can pull it off?

**Comment by Kixunil** [comment]
Awesome news: we can do it without writing any tool! `cargo fix` alone can do it, see #4052 

@TheBlueMatt @afilini you're probably the largest stakeholders, does this seem acceptable to you? I'm willing to make the PRs to your crates to patch these up once we get this released. It takes just a few minutes.

**Comment by Kixunil** [comment]
Why though? I understand not using `cargo fix` in general but in this case it's a single change, provided the strategy I outlined above is used. Nothing can go wrong basically and even then you can just skim through the code to double check and run the test suite.

**Comment by apoelstra** [comment]
How do you use `cargo fix` on just this one specific breakage? If we release a new version there'll be lots of API stuff.

**Comment by Kixunil** [comment]
That's what I wrote above. Make a pre-release with all breaking changes except this one and then the real release also contains the rename. People can temporarily upgrade to the pre-release (just in git), fix the other stuff, commit, then upgrade to the real 1.0, run `cargo fix`. They could've done it with `[patch]` too but this will be nicer and not a huge deal for us.

**Comment by dpc** [comment]
BTW.[ For unrelated reasons, I'd love Rust to have a `rename`-ability](https://internals.rust-lang.org/t/could-we-ever-fix-naming-of-io-write/22226), so that one can rename a symbol, marking old one as deprecated, but still working and basically being an alias. This would allow much smoother evolution of Rust code.

**Comment by Kixunil** [comment]
@dpc yes, that'd be superior but in this case we want an attribute saying "this rename is just about the name, no semantics change". Many (most?) renames are also semantics changes. Like the one you mentioned (not that the semantics of the function changed but "if you used `write` before you should double check your code; also I agree with you that doing something about it would be great).

**Comment by Kixunil** [comment]
There's a difference. In case of singular/plural or some typos we already had it doesn't prevent mistakes, just confused looks into doc etc. But in case of `write` to `partial_write` ideally everyone would inspect all cases it finds and check if they are correct. Otherwise it'll only prevent mistakes in the future but does nothing about the current ones. To be clear, I want both to exist. "just rename" should allow `cargo` fix to blindly accept it while "rename with semantic implications" should not.

**Comment by dpc** [comment]
I'm a bit off on your point, possibly because I didn't follow the conversation. All I'm saying is: I'd love to have an ability in Rust to do renames without semantic changes, for the purposes of polishing names without breaking semver, which is actually quite common. After all that's one of the hardest problems in comp. science.

**Comment by apoelstra** [comment]
> Make a pre-release with all breaking changes except this one

Lol there is no way I am going to update to a specific (series of?) pre-release with the goal of running a `cargo fix` command. That is even harder and more annoying than just fixing my code with `sed`, plus it won't be discoverable.

I don't believe our users are any more tolerant of this kind of thing than me.

**Comment by Kixunil** [comment]
> That is even harder and more annoying than just fixing my code with `sed`

How? It's just one version bump in `Cargo.toml` (or at worst a few of them in a large project). `cargo fix` works semantically, it's much more reliable than `sed`. If it didn't have the edge case I mentioned (which is in a way good thing it has but also maybe fied in newer Rust - I tested on MSRV only so far) it'd be perfect.

> plus it won't be discoverable.

I had a super large campaign in mind similar to that one when we made a shitload of changes and wrote an upgrade guide. And maybe make eve n bigger on this time.
